File: src/main/java/org/kite9/diagram/adl/Rectangular.java
package org.kite9.diagram.adl;

import org.kite9.diagram.position.Direction;
import org.kite9.diagram.position.RectangleRenderingInformation;
import org.kite9.diagram.style.DiagramElementSizing;

/**
 * Marker interface for diagram elements which consume a rectangular space, and therefore
 * return {@link RectangleRenderingInformation}.
 * 
 * @author robmoffat
 *
 */
public interface Rectangular extends DiagramElement {

	public RectangleRenderingInformation getRenderingInformation();
	
	/**
	 * Margin is the minimum distance from this element to elements around it that it's not connected to.
	 */
	public double getMargin(Direction d);
	
	/**
	 * Padding is the space inside this element that is consumed over and above the space of the child
	 * elements.
	 */
	public double getPadding(Direction d);
	
	public DiagramElementSizing getSizing();
}

File: kite9-visualization/src/common/kotlin/org/kite9/diagram/model/Rectangular.kt
package org.kite9.diagram.model

import org.kite9.diagram.model.position.RectangleRenderingInformation
import org.kite9.diagram.model.style.ContainerPosition

/**
 * Marker interface for diagram elements which consume a rectangular space, and therefore
 * return [RectangleRenderingInformation].
 *
 * @author robmoffat
 */
interface Rectangular : DiagramElement {

    override fun getRenderingInformation(): RectangleRenderingInformation

    /**
     * Any other details about how this rectangular is to be positioned in the container.
     */
    fun getContainerPosition(): ContainerPosition?
}